test_pull_requests

```diff
@@ -461,31 +413,107 @@ class FlushableStore
   end
 end
 
-# Quick-and-dirty middleware that attempts to invalidate appropriate
-# cache entries before handing the request off to the
-# Faraday::HttpCache middleware
-class GitHubCacheInvalidator < Faraday::Middleware
 
-  UNSAFE_METHODS = [:post, :put, :delete, :patch].freeze
+class InvalidatingGitHubCache < Faraday::HttpCache
+
+  class CustomInvalidatingStorage < Faraday::HttpCache::Storage
+
+    attr_accessor :invalidatable
+
+    def initialize(*args)
+      super
+      @invalidatable = Hash.new { |h,k| h[k] = {} }
+      @cache.storage = self
+    end
+
+    # GitHub-specific (although probably other APIs are similar) method
+    # to remove the pagination parameters from a URL so it can be better
+    # classified
+    #
+    # Example:
+    #   synthesize_url("https://api.github.com/some/resource/id?per_page=100")
+    #     returns
+    #   "https://api.github.com/some/resource/id"
+    def synthesize_url(orig_url)
+      url = orig_url.dup
+      uri = Addressable::URI.parse url
+      params = uri.query_values
+      if params
+        params.delete 'per_page'
+        params.delete 'page'
+        uri.query_values = params
+        url = uri.to_s.chomp('?')
+        # get rid of trailing numeric ID in REST path;
+        # e.g. '/issues/comments/311375581' -> '/issues/comments'
+        url.gsub!(/\/[0-9]+$/, '')
+        # get rid of trailing SHA ID in REST path
+        url.gsub!(/\/[0-9a-f]{40}$/, '')
+      end
+      url
+    end
+
+    # On write, classify the synthesized URL for later invalidation
+    def write(request, response)
+      super
+      key = cache_key_for(request.url)
+      url = synthesize_url(request.url)
+      @invalidatable[url.to_s][key] = true
+    end
+
+    # We invalidate more than is strictly necessary to handle e.g.:
+    #  DELETE https://api.github.com/repos/user/repo/issues/comments/111111111
+    #    should invalidate: https://api.github.com/repos/user/repo/issues/1/comments?per_page=100
+    def invalidate_urls(url=nil)
+      iurls=[]
+      if url
+        base_url = synthesize_url(url)
+        iurls = @invalidatable.keys.select { |iurl| iurl.to_s.start_with? base_url.to_s }
+      else
+        iurls = @invalidatable.keys
+      end
+      iurls.each do |iurl|
+        cache_keys = @invalidatable.delete(iurl)
+        cache_keys.keys.each do |cache_key|
```

Comment:
I would break out the body of this loop into a method and give it a name -- can't figure out what we're doing here from a quick glance